### PR TITLE
feat(FEC-11300): dual screen provider changes

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

Add custom data api to enable plugins custom BE calls

solves FEC-11300

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [X] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/k-provider/ovp/loaders/custom-data-loader.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+   * @param {Object} params loader params
+   */
+  constructor(params: Object) {
+    const config = OVPConfiguration.get();
+    this.requests = params.requests.map(request => ({
+      ...request,
+      url: request.getUrl(config.serviceUrl),
+      params: {...request.params, ks: params.ks}
+    }));
+    this._pluginName = params.pluginName;
```
Comment (yairans) [inline src/k-provider/ovp/loaders/custom-data-loader.js:25]: remove

File: src/k-provider/ovp/loaders/custom-data-loader.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  set requests(requests: Array<RequestBuilder>) {
+    this._requests = requests;
+  }
+
+  get requests(): Array<RequestBuilder> {
+    return this._requests;
+  }
+
+  get pluginName(): string {
```
Comment (yairans) [inline src/k-provider/ovp/loaders/custom-data-loader.js:36]: remove

File: src/k-provider/ovp/provider.js [hunk 3]
```diff
@@ -84,6 +104,20 @@ export default class OVPProvider extends BaseProvider<OVPProviderMediaInfoObject
     }
   }
 
+  _parseCustomDataFromResponse(data: Map<string, Function>): any {
+    this._logger.debug('Custom Data parsing started');
+    let response = {};
+    if (data) {
+      if (data.has(OVPCustomDataLoader.id)) {
```
Comment (yairans) [inline src/k-provider/ovp/provider.js:111]: move to the upper if

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/k-provider/ovp/loaders/custom-data-loader.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+   * @constructor
+   * @param {Object} params loader params
+   */
+  constructor(params: Object) {
+    const config = OVPConfiguration.get();
+    this.requests = params.requests.map(request => ({
+      ...request,
+      url: request.getUrl(config.serviceUrl),
+      params: {...request.params, ks: params.ks}
+    }));
```
Comment (OrenMe) [inline src/k-provider/ovp/loaders/custom-data-loader.js:24]: seems like we have a convention for this to be set inside a method called `buildRequests` so lets keep it

**Review by OrenMe — DISMISSED** [review]
would love to hear your thoughts, but for me it makes a lot more sense that each plugin creates its own loader suited to its own needs.
The beef of the loader is in the buildRequests and set response which as can be seen in all other loaders contain specific implementation details of that loader.
In this case it seems this is not a custom data loader - but a generic data loader and in any case the plugin will add on top of this its own set response logic and also parsing of response.

WDYT? what is the argument of creating the custom data loader? also pay attention that you are extending the ILoader (which can be argued is not the best interface) and adding plugin context to it - one more for managing it in the plugin.
The only thing I see is that there is some boilerplate for a loader, and for that you can maybe create a factory function that gets only buildRequests and response setter and return you a loader you can use in the `getCustomData` which I will change its name as well.

File: src/k-provider/ovp/index.js [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
+import {RESPONSE_PROFILE_TYPE} from './request-params/base-entry-response-profile';
 
 declare var __VERSION__: string;
 declare var __NAME__: string;
 
 const NAME = __NAME__ + '-ovp';
 const VERSION = __VERSION__;
 
-export {OVPProvider as Provider, NAME, VERSION};
+export {OVPProvider as Provider, NAME, VERSION, RequestBuilder, RESPONSE_PROFILE_TYPE as ResponseProfileType};
```
Comment (yairans) [inline src/k-provider/ovp/index.js:12]: in second thought we can with no exporting the response profile type as this is not provider enum but server enum.
also, we might want to export all provider enums under Enum object or something like this. so I don't want to keep this was for backward compatibility.

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/k-provider/ovp/loaders/custom-data-loader.js (hunk 2 above)
Comment (RoyBregman) [reply src/k-provider/ovp/loaders/custom-data-loader.js]: removed

File: src/k-provider/ovp/loaders/custom-data-loader.js (hunk 1 above)
Comment (RoyBregman) [reply src/k-provider/ovp/loaders/custom-data-loader.js]: removed

File: src/k-provider/ovp/loaders/custom-data-loader.js (hunk 4 above)
Comment (RoyBregman) [reply src/k-provider/ovp/loaders/custom-data-loader.js]: Not relevant anymore as the entire Loader was taken out the providers and moved to dual screen

File: flow-typed/types/external-request-loader.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// @flow
+declare type ExternalRequestLoader = {
```
Comment (OrenMe) [inline flow-typed/types/external-request-loader.js:2]: Not sure we need to call this `external`, simply request loader, as this is correct for both internal and external

File: src/k-provider/ovp/index.js [hunk 7]
```diff
@@ -1,10 +1,14 @@
...
 declare var __VERSION__: string;
 declare var __NAME__: string;
 
 const NAME = __NAME__ + '-ovp';
 const VERSION = __VERSION__;
 
-export {OVPProvider as Provider, NAME, VERSION};
+const ENUMS = {RESPONSE_PROFILE_TYPE};
+
+export {OVPProvider as Provider, NAME, VERSION, RequestBuilder, ENUMS};
```
Comment (RoyBregman) [inline src/k-provider/ovp/index.js:14]: remove exporting enums. Need to rethink about export. For now will be used locally in the dual screen plugin

File: src/k-provider/ovp/provider.js [hunk 8]
```diff
@@ -70,6 +70,27 @@ export default class OVPProvider extends BaseProvider<OVPProviderMediaInfoObject
     });
   }
 
+  doRequest(loaders: Array<ExternalRequestLoader>, params: any = {}): Promise<any> {
```
Comment (RoyBregman) [inline src/k-provider/ovp/provider.js:73]: add docs / tests / guides

File: src/k-provider/ovp/provider.js [hunk 9]
```diff
@@ -70,6 +70,27 @@ export default class OVPProvider extends BaseProvider<OVPProviderMediaInfoObject
...
+    const dataLoader = new OVPDataLoaderManager(this.playerVersion, this.partnerId, params.ks || this.ks, this._networkRetryConfig);
```
Comment (RoyBregman) [inline src/k-provider/ovp/provider.js:74]: Do we need to pass the KS? Can't we rely on the ks in the provider?

File: flow-typed/types/external-request-loader.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+  loader: Function,
```
Comment (yairans) [inline flow-typed/types/external-request-loader.js:3]:
```suggestion
  loader: ILoader,
```

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/k-provider/ovp/provider.js [hunk 11]
```diff
@@ -70,6 +70,27 @@ export default class OVPProvider extends BaseProvider<OVPProviderMediaInfoObject
...
+
+    return new Promise((resolve, reject) => {
+      loaders.forEach((loaderRequest: ExternalRequestLoader) => {
+        dataLoader.add(loaderRequest.loader, loaderRequest.params);
```
Comment (yairans) [inline src/k-provider/ovp/provider.js:78]: loader must include id so add it to the ILoader interface

**Review by yairans — CHANGES_REQUESTED** [review]

File: flow-typed/types/external-request-loader.js (hunk 6 above)
Comment (RoyBregman) [reply flow-typed/types/external-request-loader.js:2]: done

File: src/k-provider/ovp/index.js (hunk 7 above)
Comment (RoyBregman) [reply src/k-provider/ovp/index.js]: done

File: src/k-provider/ovp/provider.js (hunk 8 above)
Comment (RoyBregman) [reply src/k-provider/ovp/provider.js]: opened a separate ticket - https://kaltura.atlassian.net/browse/FEC-11336

File: src/k-provider/ovp/provider.js (hunk 11 above)
Comment (RoyBregman) [reply src/k-provider/ovp/provider.js:78]: Unfortunately it is not possible to make status properties in an Interface. Will be documented instead

File: src/k-provider/ovp/provider.js (hunk 9 above)
Comment (RoyBregman) [reply src/k-provider/ovp/provider.js]: As agreed we will use the ks from the provider and if needed the external call will add the ks to the specific service request

File: src/k-provider/ovp/index.js (hunk 5 above)
Comment (RoyBregman) [reply src/k-provider/ovp/index.js]: done

File: src/k-provider/ovp/provider.js (hunk 3 above)
Comment (RoyBregman) [reply src/k-provider/ovp/provider.js]: Not relevant anymore

**Review by yairans — APPROVED** [review]

**Pull request merged**